src/ngraph/node.hpp

```diff
@@ -90,11 +90,8 @@ namespace ngraph
                 input.get_output().remove_input(&input);
             }
         }
-        virtual void generate_adjoints(autodiff::Adjoints& adjoints,
-                                       const std::shared_ptr<Node>& delta)
-        {
-        }
 
+        virtual void generate_adjoints(autodiff::Adjoints& adjoints, const NodeVector& deltas) {}
```

Comment:
It seems you only ever use the first value in these vectors?

Reply:
so far, yes. Even `BatchNorm` only uses `deltas.at(0)` since the first delta corresponds to `dinput`. It could have easily been any `deltas.at(1)` or `deltas.at(2)` depending an order of the outputs of `BatchNorm`. We can now handle all these scenarios.

https://github.com/NervanaSystems/ngraph/pull/796/files#diff-0540104e61a41e6ff57c2cee5830d723L260

Also, note the implementation of `GetOutputElement::generate_adjoints`

https://github.com/NervanaSystems/ngraph/pull/796/files#diff-0065dc21e94515d9e5b9dd2f63a1021dR54

It pushes its update to the corresponding output.